hikari/utilities/event_stream.py

```diff
@@ -0,0 +1,249 @@
+        # We accept `None` to represent unlimited here to be consistent with how `None` is already used to represent
+        # unlimited for timeout in other places.
+        self._queue: asyncio.Queue[EventT] = asyncio.Queue(limit or 0)
+        self._registered_listener: typing.Optional[
+            typing.Callable[[EventT], typing.Coroutine[typing.Any, typing.Any, None]]
+        ] = None
+        # The registered wrapping function for the weak ref to this class's _listener method.
+        self._timeout = timeout
+
+    async def _listener(self, event: EventT) -> None:
+        if not self._filters(event):
+            return
+
+        try:
+            self._queue.put_nowait(event)
+        except asyncio.QueueFull:
+            pass
+
+    async def __anext__(self) -> EventT:
+        if not self._active:
+            raise TypeError("stream must be started with `await with` before entering it")
+
+        try:
+            return await asyncio.wait_for(self._queue.get(), timeout=self._timeout)
+        except asyncio.TimeoutError:
+            raise StopAsyncIteration from None
+
+    async def _await_all(self) -> typing.Sequence[EventT]:
+        await self.open()
+        result = [event async for event in self]
+        await self.close()
+        return result
+
+    def __await__(self) -> typing.Generator[None, None, typing.Sequence[EventT]]:
+        return self._await_all().__await__()
+
+    def __del__(self) -> None:
+        # For the sake of protecting highly intelligent people who forget to close this, we register the event listener
+        # with a weakref then try to close this on deletion. While this may lead to their consoles being spammed, this
+        # is a small price to pay as it'll be way more obvious what's wrong than if we just left them with a vague
+        # ominous memory leak.
+        if self._active:
+            _LOGGER.warning("active %r streamer fell out of scope before being closed", self._event_type.__name__)
+            asyncio.ensure_future(self.close())
+
+    async def close(self) -> None:
+        if self._active and self._registered_listener is not None:
+            try:
+                self._app.dispatcher.unsubscribe(self._event_type, self._registered_listener)
+            except ValueError:
+                pass
+
+            self._registered_listener = None
+
+        self._active = False
+
+    def filter(
+        self,
+        *predicates: typing.Union[typing.Tuple[str, typing.Any], typing.Callable[[EventT], bool]],
+        **attrs: typing.Any,
+    ) -> typing.Union[EventStream[EventT], iterators.LazyIterator[EventT]]:
+        if self._active:
+            return super().filter(*predicates, **attrs)
+
+        new_conditions = self._map_predicates_and_attr_getters("filter", *predicates, **attrs)
+        if self._filters:
+            self._filters |= new_conditions
+        else:
+            self._filters = new_conditions
```

Comment:
is this reachable? Wont line 233 suffice?

Reply:
it is reachable as i added a __bool__ override to All which returns False if no conditions are set else True but line 233 would also suffice so eh